python/servo/devtools_tests.py

```diff
@@ -43,28 +44,96 @@ class Devtools:
     client: RDPClient
     watcher: WatcherActor
     targets: list
+    exited: bool = False
+
+    def connect(*, expected_targets=1):
+        """
+        Connect to the Servo devtools server.
+        You should use a `with` statement to ensure we disconnect unconditionally.
+        """
+        client = RDPClient()
+        client.connect("127.0.0.1", 6080)
+        root = RootActor(client)
+        tabs = root.list_tabs()
+        tab_dict = tabs[0]
+        tab = TabActor(client, tab_dict["actor"])
+        watcher = tab.get_watcher()
+        watcher = WatcherActor(client, watcher["actor"])
+
+        done = Future()
+        targets = []
+
+        def on_target(data):
+            try:
+                targets.append(data["target"])
+                if len(targets) == expected_targets:
+                    done.set_result(None)
+            except Exception as e:
+                # Raising here does nothing, for some reason.
+                # Send the exception back so it can be raised.
+                done.set_result(e)
+
+        client.add_event_listener(
+            watcher.actor_id,
+            Events.Watcher.TARGET_AVAILABLE_FORM,
+            on_target,
+        )
+        watcher.watch_targets(WatcherActor.Targets.FRAME)
+        watcher.watch_targets(WatcherActor.Targets.WORKER)
+
+        result: Optional[Exception] = done.result(1)
+        if result:
+            raise result
+
+        return Devtools(client, watcher, targets)
+
+    def __getattribute__(self, name):
+        """
+        Access a property, raising a ValueError if the instance was previously marked as exited.
+        """
+        if name != "exited" and object.__getattribute__(self, "exited"):
+            raise ValueError("Devtools instance must not be used after __exit__()")
+        return object.__getattribute__(self, name)
+
+    def __enter__(self):
+        """
+        Enter the `with` context for this instance, raising a ValueError if it was previously marked as exited.
+        """
+        if self.exited:
+            raise ValueError("Devtools instance must not be used after __exit__()")
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```

Comment:
Please add a type here.

Reply:
done in a84c42cd8d41a0a8100cea0fd9b1af29429329fb